hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -489,13 +489,29 @@ public EntityEntry addEntry(
 
 		final EntityEntry e;
 
-		if( (entity instanceof ManagedEntity) &&  ((ManagedEntity) entity).$$_hibernate_getEntityEntry() != null && status == Status.READ_ONLY) {
-			e = ((ManagedEntity) entity).$$_hibernate_getEntityEntry();
-			e.setStatus( status  );
+		/*
+		Important: 		the following instanceof checks and castings are intentional.
+						Please do *not* refactor to make calls through the EntityEntryFactory
+						interface. Refactoring will results in polymorphic call sites which will
+						severely impact performance.
+		*/
+		if (persister.getEntityEntryFactory() instanceof MutableEntityEntryFactory){
+			e = ((MutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
+					status,
+					loadedState,
+					rowId,
+					id,
+					version,
+					lockMode,
+					existsInDatabase,
+					persister,
+					disableVersionIncrement,
+					lazyPropertiesAreUnfetched,
+					this
+			);
 		}
 		else {
-			final EntityEntryFactory entityEntryFactory = persister.getEntityEntryFactory();
-			e = entityEntryFactory.createEntityEntry(
+			e = ((ImmutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
 					status,
```

Comment:
We are casting the interface to the known implementation to prevent the JVM from having to resolve which implementations method .createEntityEntry() to call.  This removes the need make calls via polymorphic call sites and reverts the perf regression introduced by having more than one implementation of the EntityEntryFactory interface.